rust/cymbal/src/assignment_rules.rs

```diff
@@ -98,36 +150,67 @@ impl AssignmentRule {
         Ok(())
     }
 
-    pub async fn apply<'c, E>(&self, conn: E, issue_id: Uuid) -> Result<Assignment, sqlx::Error>
-    where
-        E: sqlx::Executor<'c, Database = sqlx::Postgres>,
-    {
-        // TODO - should we respect existing assignments? This does, and I think that's right, but :shrug:
-        let assignment = sqlx::query_as!(
-            Assignment,
-            r#"
-                INSERT INTO posthog_errortrackingissueassignment (id, issue_id, user_id, user_group_id, role_id, created_at)
-                VALUES ($1, $2, $3, $4, $5, NOW())
-                ON CONFLICT (issue_id) DO UPDATE SET issue_id = $2 -- no-op to get a returned row
-                RETURNING id, issue_id, user_id, user_group_id, role_id, created_at
-            "#,
-            Uuid::now_v7(),
-            issue_id,
-            self.user_id,
-            self.user_group_id,
-            self.role_id,
-        ).fetch_one(conn).await?;
+    pub fn try_match(
+        &self,
+        issue: &Value,
+        props: &Value,
+    ) -> Result<Option<NewAssignment>, VmError> {
+        let rule_bytecode = match &self.bytecode {
+            Value::Array(ops) => ops,
+            _ => {
+                return Err(VmError::Other(format!(
+                    "Invalid rule bytecode - expected array, got {:?}",
+                    self.bytecode
+                )))
+            }
+        };
 
-        Ok(assignment)
+        let mut globals = HashMap::new();
+        globals.insert("issue".to_string(), issue.clone());
+        globals.insert("properties".to_string(), props.clone());
+        let globals: Value = serde_json::to_value(globals)
+            .expect("Can construct a json object from a hashmap of String:JsonValue");
+        let program = Program::new(rule_bytecode.clone())?;
+        let context = ExecutionContext::with_defaults(program).with_globals(globals);
+        let mut vm = context.to_vm()?;
+
+        metrics::counter!(ASSIGNMENT_RULES_TRIED).increment(1);
+
+        let mut i = 0;
+        while i < context.max_steps {
+            let step_result = vm.step()?;
+            match step_result {
+                StepOutcome::Finished(Value::Bool(b)) if b => {
+                    return Ok(NewAssignment::try_new(
+                        self.user_id,
+                        self.user_group_id,
+                        self.role_id,
+                    ))
+                }
+                StepOutcome::Finished(res) => {
+                    return Err(VmError::Other(format!(
+                        "Assignment rule returned {:?}, expected a boolean value",
+                        res
+                    )))
+                }
+                StepOutcome::NativeCall(name, args) => {
+                    context.execute_native_function_call(&mut vm, &name, args)?
+                }
+                StepOutcome::Continue => {}
+            }
+            i += 1;
+        }
```

Comment:
Feels like a lot of this is shared with the grouping rules code. Happy to keep them separate initially but we should be conscious in case they diverge